Port stage0 to darwin

discordianfish opened the pull request:
I cleaned up my changes *a bit* but I'm afraid I won't have time to continue on this further. Maybe we'll decide that we want this and I can invest some more time but doesn't look like it right now...

I have no experience with autotools and didn't have time to understand the fairly complex build lib stuff here. So I just removed the static libc check and ignored the 'warnings' the Makefiles are throwing.

I've rebased my changes on #2868 since we need to disable systemd-journal support.

Unfortunately these aren't the only changes necessary. The system libs attr and acl need darwin specific patches as well which I have a in branch. I won't have time to submit those upstream anytime soon either, so if someone wants to pick that up that would be great (C and especially autoconf and preprocessor magic are outside my comfort zone). The patched libs are:

- https://github.com/discordianfish/attr
- https://github.com/discordianfish/acl

Here is a complete step-by-step listing how to get this working on osx / el capitan:

```shell
brew go mercurial autoconf automake md5sha1sum squashfs wget gettext
PATH=$PATH:/usr/local/opt/gettext/bin # for gettext tools to be found

git clone -b fish/port-darwin git@github.com:discordianfish/attr.git 
cd attr
./autogen.sh
./configure --prefix ~/local
make install

cd ..
git clone -b fish/port-darwin git@github.com:discordianfish/acl.git 
cd acl
CFLAGS="-I$HOME/local/include" LDFLAGS="-L$HOME/local/lib" ./configure --prefix ~/local
make install
perl -pi -e 's/^EXPORT //' ~/local/include/{sys/acl,acl/libacl}.h 

export GOPATH=$HOME/dev/go
mkdir -p $GOPATH/src/github.com/coreos
cd $GOPATH/src/github.com/coreos
git clone -b fish/port-darwin git@github.com:discordianfish/rkt.git
cd rkt
./autogen.sh
CFLAGS="-I$HOME/local/include" LDFLAGS="-L$HOME/local/lib" ./configure --with-stage1-flavors=none --disable-tpm --enable-sdjournal=no
make
```

Make will still fail (tbh, haven't figured out if there is a target to only build stage0) but it builds rkt stage:

```
$ file build-rkt-1.10.0+git/target/bin/rkt
build-rkt-1.10.0+git/target/bin/rkt: Mach-O 64-bit executable x86_64
```

And a few things work:
```
$ sudo build-rkt-1.10.0+git/target/bin/rkt --insecure-options=image fetch docker://quay.io/prometheus/prometheus
image: remote fetching from URL "docker://quay.io/prometheus/prometheus"
Downloading sha256:a3ed95caeb0 [=============================]       32 B / 32 B
Downloading sha256:a3ed95caeb0 [=============================]       32 B / 32 B
Downloading sha256:a3ed95caeb0 [=============================]       32 B / 32 B
Downloading sha256:318882358ff [=============================]     658 B / 658 B
Downloading sha256:a3ed95caeb0 [=============================]       32 B / 32 B
Downloading sha256:a3ed95caeb0 [=============================]       32 B / 32 B
Downloading sha256:0717a09aa2b [=============================] 6.71 MB / 6.71 MB
Downloading sha256:a3ed95caeb0 [=============================]       32 B / 32 B
Downloading sha256:a3ed95caeb0 [=============================]       32 B / 32 B
Downloading sha256:a3ed95caeb0 [=============================]       32 B / 32 B
Downloading sha256:b561633a5ad [=============================] 3.49 MB / 3.49 MB
Downloading sha256:385e281300c [=============================]   676 KB / 676 KB
Downloading sha256:27c10d20424 [=============================] 6.68 KB / 6.68 KB
Downloading sha256:66dade7adfc [=============================] 2.83 KB / 2.83 KB
Downloading sha256:e418e02f5f3 [=============================]   488 KB / 488 KB
error opening /proc/filesystems
sha512-23609a714ace62e219451e77b08ae19f
```

Partial implementation of #2797

lucab commented:
ok to test

lucab commented:
@discordianfish thanks for your work on this. This PR contains a lot of unrelated stuff so it is clearly not mergeable as is, but it traced the path and it is a good starting point to build upon. I will try to scavenge the useful bits in it to make things more portable and split them into self-contained PRs.

discordianfish commented:
@lucab Yes, it's based on #2868. Not sure tbh why it's including all those other commits which I suppose are already in master. Probably best to get #2868 merged after which I can rebase on master? Or cherry-pick that commit?

lucab commented:
It depends on your time availability, but yes it would be nice if you could rebase on #2868 (just merged).

discordianfish commented:
@lucab doesn't look like it's merged yet. Will rebase afterwards.

discordianfish commented:
@lucab It's rebased although I don't know why the semaphoreci stuff failed. Doesn't look related.

discordianfish commented:
@lucab I assume nobody from your side can have a look at the autotool issues? I can look into submitting the upstream changes soon

jonboulle commented:
@discordianfish what do you need help with specifically w.r.t autotools?

discordianfish commented:
@jonboulle In general making it stop throwing warnings.. More specifically:

- realpath not found and I can't find a reference to it
- make it possible to skip building stage1 (right now if you run make, it will build stage 0 just fine but error out when building stage 1)

I assume for somebody familiar with the autotools setup here it's a matter a minutes.

I've just rebased this. Let me know what needs to be done to get this merged. After that, I can look into porting stage1.

lucab commented on rkt/api_service_stub.go:
```diff
@@ -12,7 +12,7 @@
 // limitations under the License.
 
-// +build !sdjournal
+// +build !sdjournal,linux
```
I think this is a redundant combination, as I guess on Mac you need to build with `--enable-sdjournal=no` anyway. Am I correct?

discordianfish replied:
Actually no, now that I'm trying to remove it I realize that the whole api parts won't build on darwin, so it all is disabled. (If we don't disable it here it results in `rkt/api_service_stub.go:25: undefined: v1AlphaAPIServer`)

lucab commented on pkg/sys/sys_darwin.go:
```diff
@@ -0,0 +1,68 @@
+
+// not implemented
+func RemountPrivate(target string) error {
+	return nil
```
This should probably return a "Not implemented" error.

lucab commented on configure.ac:
```diff
@@ -357,33 +357,6 @@ AS_VAR_IF([RKT_SPECIFIED_FLAVORS],[],
-                                          [AC_MSG_RESULT([yes])
-                                           RKT_DEFINES_FOR_ENTER=-DNO_SETNS_AVAILABLE],
-                                          [AC_MSG_RESULT([no])
-                                           AC_MSG_ERROR([*** Neither setns function nor __NR_setns macro are available - either both glibc and kernel are too old or their development headers are not installed])])])
```
Mostly a note to self here: we this (on Linux) to check if the current glibc is recent enough to build stage1 tools. We encountered issues on Centos 6 already. This should be made conditional on the OS or on the stage1 being built.

lucab commented on pkg/pod/pods.go:
```diff
@@ -711,6 +712,62 @@ func (p *Pod) getModTime(path string) (time.Time, error) {
+	}
+	return fileutil.GetCtime(fi), nil
+}
+
```
I think most of this snippet is a leftover from a previous iteration, and you just need the `fileutil.GetCtime` change in here. Can you please double-check?

discordianfish replied:
Oh right, cleaning up..

lucab commented:
@discordianfish can you please paste somewhere logs for the errors you are seeing? I'm especially unsure about the `realpath` one, as it may be either checking for the C function or for the system utility.

Regarding a stage0-only build, you may do it already via:
```
./configure --with-stage1-default-name=coreos.com/rkt/stage1-fly --with-stage1-default-version=1.16.0
```
but then you need a proper stage1 already in store.

discordianfish commented:
@lucab Ok, I've addressed your comments and with those stage1 flags the make run finishes successfully. I've ran:
```
$ CFLAGS="-I$HOME/local/include" LDFLAGS="-L$HOME/local/lib" ./configure --with-stage1-default-name=coreos.com/rkt/stage1-fly --with-stage1-default-version=1.16.0 --disable-tpm --enable-sdjournal=no
$ make
```

Here is the full output: https://gist.github.com/discordianfish/24ddc3d386edd40a0bc13445d72fb567

Regarding the realpath issue: I realize you can install realpath on osx via brew which seem to fix the issue. This revealed another issue which I fixed in the last commit. But I still get this warning:

```
$ make
ld: unknown option: --verbose
Bad path: /libnss_files.so.2
usage: basename string [suffix]
       basename [-a] [-s suffix] string [...]
```

It looks like it's doing some preps for building stage1 which shouldn't be necessary. Not sure how to skip this though. But it's just a warning anyway, so I'd be good with merging this as it is for now :)

discordianfish commented:
Guess this will never get merged. Should I close it?
